keps/sig-instrumentation/1748-pod-resource-metrics/README.md

```diff
@@ -0,0 +1,497 @@
+
+The reservation of a pod for a given resource is
+
+    max(max_over_init_containers(resource), sum_over_containers(resource)) + resource_overhead
+
+The reservation of a pod is the union of all resource calculations of this form. Resources are considered individually when making decisions about admission or rejection from nodes or quota, and any one resource being exceeded will prevent execution or admission.
+
+For instance, the pod:
+
+```
+metadata:
+  name: nginx
+spec:
+  initContainers:
+  - name: copy-files
+    ...
+    resources:
+      requests:
+        cpu: 100m
+  - name: generate-config
+    ...
+    resources:
+      requests:
+        cpu: 300m
+  containers:
+  - name: proxy
+    ...
+    resources:
+      requests:
+        cpu: 50m
+  - name: sidecar-logging
+    ...
+    resources:
+      requests:
+        cpu: 100m
+```
+
+will be handled by the system in the following ways:
+
+* Is considered to consume `300m` cores of CPU by the scheduler (the node must be able to run the largest init container by itself, which is larger than the sum of the proxy and sidecar containers at runtime)
+* The kubelet will not admit the pod if `(the node's allocatable CPU) - (the sum of all running pods)` is less than `300m`
+* The kubelet will create a cgroup for the pod that expects to get roughly `300m` cores, but the container cgroup created for `copy-files` will request `100m` while that init container is running
+* The quota subsystem will block this pod from being created if there is less than `300m` available.
+
+Once a pod is created, it passes through four high level lifecycle phases as seen by the total Kubernetes system. The first phase is `Pending` - the time before the pod is scheduled to a node. The second phase is `Initializing` - the time between when the pod is scheduled on a node and when all init containers have completed successfully and the `Initalized` condition on the pod is set to `true`. A pod without init containers may only remain in this phase very briefly. The third phase is `Running`, when all init containers have completed, and continues until the pod is deleted or reaches a terminal state of success or failure.  The final state is `Completed` which means the pod has no running containers, all resources have been released or cleaned up, and the pod will never again consume those resources.
+
+Note that Kubernetes is a distributed system and the kubelet owns the pod lifecycle, so the value reported by the system will always slightly trail the actual state of the system.  The lifecycle metric may report a pod is `Pending` but some or all resources may already be allocated on the node, or a `Running` pod may have already reached completion but the Kubelet has not yet reported that to the apiserver.
```

Comment:
If we don't already then maybe current resources allocated by running pods on a kubelet should be instrumented as well.

Reply:
So that's an interesting point - I kind of assumed that since the API is truth, the API representation as seen by both kubelet, quota, and scheduler is actually authoritative.  So it is sufficient to calculate the resource model as seen by the API rather than duplicating it on the kubelet, since the kubelet already owns the responsibility of accurately keeping the API up to date (vs creating a separate mechanism).

That said, having totals per kubelet might be interesting from a "compare expected vs actual" perspective, but it's not even nearly as interesting as the summarization of the total model.

Reply:
> That said, having totals per kubelet might be interesting from a "compare expected vs actual" perspective, but it's not even nearly as interesting as the summarization of the total model.

Agreed. This was my thought as well about expected vs actual. It's entirely additional and more for troubleshooting than the premise of the KEP.